### PR TITLE
TimeHelper: migrate to Java 8's `java.time` #8362

### LINKED ISSUE #8632: Help: Not able to login in development environment.
Ubuntu 16.04.4 LTS
Java version "1.8.0_161"
Node version v9.6.1
Chrome 64.0.3282.186

I am able to successfully run the development environment but when I open the localhost I am unable to login as student or instructor.

This is what I get when I try to log in as a student:-
screenshot from 2018-03-13 18-41-54

This is what I get when I try to log in as an instructor:-
screenshot from 2018-03-13 18-42-03

### PR DESCRIPTION
**Pull request by tran-tien-dat**
Part of #8362 

**Outline of Solution**

This is a first attempt at migrating part of our whole codebase to `java.time` classes, focusing on only 1 function `TimeHelper.combineDateTime`. Its logic has also been simplified.

The method now returns a `java.time.LocalDateTime`, but it is converted back to `java.util.Date` immediately wherever the method is called. This is to keep the changes localized, and easier to review.

Also, I want to discuss whether this should be the right way to carry out this migration.

### CODE REVIEW

**Comment by tran-tien-dat** [comment]
@darrenwee

File: src/main/java/teammates/common/util/TimeHelper.java [hunk 1]
```diff
@@ -427,31 +434,6 @@ public static double getLocalTimezoneHourOffset() {
         return TimeZone.getDefault().getOffset(new Date().getTime()) / 1000.0 / 60.0 / 60.0;
     }
 
-    private static Date convertToDate(String date, int time) {
```
Comment (tran-tien-dat) [inline src/main/java/teammates/common/util/TimeHelper.java]: This method is only used once in `combineDateTime`, so I have merged it there.

File: src/test/java/teammates/test/cases/datatransfer/FeedbackSessionAttributesTest.java [hunk 2]
```diff
@@ -26,8 +26,10 @@
 
     @BeforeClass
     public void classSetup() {
-        startTime = TimeHelper.combineDateTime("09/05/2016", "1000");
```
Comment (tran-tien-dat) [inline src/test/java/teammates/test/cases/datatransfer/FeedbackSessionAttributesTest.java]: Note that this `"1000"` is a mistake, since `combineDateTime` is supposed to take in the hour only, not the minute

File: src/test/java/teammates/test/cases/util/TimeHelperTest.java [hunk 3]
```diff
@@ -40,33 +41,22 @@ public void testFormatTimeForEvaluation() {
...
-        cal.set(2013, 1, 1, 23, 59);
-        expectedOutput = cal.getTime();
+        expectedOutput = LocalDateTime.of(2013, 2, 1, 23, 59);
         assertEquals(expectedOutput, TimeHelper.combineDateTime(testDate, testTime));
 
         ______TS("negative time");
-        cal.clear();
-        cal.set(2013, 1, 1, -5, 0);
-        expectedOutput = cal.getTime();
-        assertEquals(expectedOutput, TimeHelper.combineDateTime(testDate, "-5"));
```
Comment (tran-tien-dat) [inline src/test/java/teammates/test/cases/util/TimeHelperTest.java]: Accepting negative time or >24 hour is a weird feature of `Calendar`. Anyway, we don't need such feature and the new classes don't support them so I just simply consider it as errors

File: src/main/java/teammates/common/util/TimeHelper.java [hunk 4]
```diff
@@ -112,25 +116,30 @@ public static Calendar now(double timeZone) {
     }
 
     /**
-     * Convert a date string and time string into a Date object. Returns null on error.
-     *
-     * @param inputDate
+     * Convert a date string and time string of only the hour into a Date object.
+     * If the hour is 24, it is converted to 23:59. Returns null on error.
+     *  @param inputDate
```
Comment (amarlearning) [inline src/main/java/teammates/common/util/TimeHelper.java]: revert this change.

Comment (tran-tien-dat) [reply src/main/java/teammates/common/util/TimeHelper.java]: I am not changing the behavior of this function. I am simply adding the undocumented behavior to the javadoc

Comment (amarlearning) [reply src/main/java/teammates/common/util/TimeHelper.java]:
```diff
+ *  @param inputDate
```
I was talking about this extra space that you gave before `@param`

Comment (tran-tien-dat) [reply src/main/java/teammates/common/util/TimeHelper.java]: I see 😅, missed out that small space. Thank you for pointing it out.

**Comment by sukanta-27** [comment]
The build error you are facing has been addressed in #8414. If your master branch is out of date, update it.

**Comment by tran-tien-dat** [comment]
Yep. Thanks @sukanta-27 for the suggestion!

**Comment by tran-tien-dat** [comment]
@whipermr5 can you update the `java-time-migration` branch to the latest `master` so that my PR doesn't appear as if it's trying to change 3k lines 😄 Thanks!

File: src/main/java/teammates/common/util/TimeHelper.java [hunk 5]
```diff
@@ -425,29 +440,18 @@ public static double getLocalTimezoneHourOffset() {
...
+     */
+    public static ZoneId convertToZoneId(double timeZone) {
+        return ZoneId.ofOffset("UTC", ZoneOffset.ofTotalSeconds((int) (timeZone * 60 * 60)));
+    }
 
+    /**
+     * Temporary method for transition from java.util.Date.
+     */
+    public static Date convertLocalDateTimeToDate(LocalDateTime localDateTime) {
+        return localDateTime == null ? null : Date.from(localDateTime.atZone(ZoneId.of("UTC")).toInstant());
```
[the remark this answers is not in the document]
Comment (tran-tien-dat) [reply src/main/java/teammates/common/util/TimeHelper.java]: I will just write a comment. This is a temporary method so I don't think we need a javadoc?

[the remark this answers is not in the document]
Comment (darrenwee) [reply src/main/java/teammates/common/util/TimeHelper.java]: You could just write `@param localDateTime the local datetime in UTC` and it'd be more convenient since intellij will show it in the quick doc and we don't have to remember that it must be in UTC.

File: src/main/java/teammates/common/util/TimeHelper.java [hunk 6]
```diff
@@ -112,25 +116,36 @@ public static Calendar now(double timeZone) {
...
-     *
+     * Convert a date string and time string of only the hour into a Date object.
+     * If the hour is 24, it is converted to 23:59. Returns null on error.
      * @param inputDate
      *            The date in format dd/MM/yyyy
      * @param inputTimeHours
-     *            The time as number of hours
+     *            0-24
      */
+    @Deprecated
```
[the remark this answers is not in the document]
Comment (tran-tien-dat) [reply src/main/java/teammates/common/util/TimeHelper.java]: It's gonna be deleted anyway and we are the only people using it so I think that's not necessary.

File: src/main/java/teammates/common/util/TimeHelper.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (darrenwee) [reply src/main/java/teammates/common/util/TimeHelper.java]: Alternatively, rename `convertLocalDateTimeToDate` to `convertUtcLocalDateTimeToDate` for self-documenting code.

[the remark this answers is not in the document]
Comment (tran-tien-dat) [reply src/main/java/teammates/common/util/TimeHelper.java]: I see your point. Ok

[the remark this answers is not in the document]
Comment (tran-tien-dat) [reply src/main/java/teammates/common/util/TimeHelper.java]: But note that if you need to think about whether when using this method, the `localDateTime` is in UTC or not, then you are probably doing the wrong thing. Timezone conversion should use other methods. This method is for blindly wrapping a `LocalDateTime` returned in the new functions that use Java 8.

File: src/main/java/teammates/common/util/TimeHelper.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (darrenwee) [reply src/main/java/teammates/common/util/TimeHelper.java]: It's just for good practice with the `@Deprecated` annotation. It's up to you!

File: src/main/java/teammates/common/util/TimeHelper.java [hunk 7]
```diff
@@ -112,25 +117,44 @@ public static Calendar now(double timeZone) {
...
+        } else {
+            dateTimeString = inputDate + " " + inputTimeHours + ".00";
+        }
         try {
-            inputTimeInt = Integer.parseInt(inputTimeHours) * 100;
-        } catch (NumberFormatException nfe) {
+            return LocalDateTime.parse(dateTimeString, formatter);
+        } catch (DateTimeParseException e) {
+            log.warning("Failed to parse datetime string: \""
+                    + dateTimeString + "\" with format " + formatter);
```
Comment (whipermr5) [inline src/main/java/teammates/common/util/TimeHelper.java]: Not sure we actually need to log this - if it happens, it's because the user sent custom POST parameters manually, which we are handling fine. The original code didn't log it after all.

**Comment by whipermr5** [comment]
@damithc For merging into a long-lived branch, do I need to request for your approval?

**Comment by damithc** [comment]
> @damithc For merging into a long-lived branch, do I need to request for your approval?

No need.
